Add timer API

pineapplemachine opened the pull request:
Add a `timer.zig` module which exposes an API for setting and reading timer-related registers.

Disclaimer: I have basically no idea what I'm doing. Hopefully this is sensible?

Based on the documentation here: https://gbadev.net/tonc/timers.html

(This PR was based on the same `builder.zig` changes as in #24, which were needed for me to build ziggba with a recent version of zig.)

pineapplemachine committed:
Update builder.zig to no longer use deprecated/recently removed API

per https://github.com/ziglang/zig/commit/3d393dba6fc06cea3508aaa3db9d49042663367e

pineapplemachine committed:
Add API for accessing timer-related registers, plus example

vortexofdoom commented on examples/secondsTimer/secondsTimer.zig:
```diff
@@ -0,0 +1,108 @@
+    // which is the same as once per second.
+    // When it oveflows, Timer 2 will be incremented by 1 due
+    // to its "cascade" flag.
+    timer.data_1.* = @truncate(-0x4000);
+    timer.ctrl_1.* = .{
+        .freq = .cycles_1024,
+        .enable = .enable,
+    };
+    timer.ctrl_2.* = .{
+        .freq = .cycles_1024,
+        .cascade = .enable,
+        .enable = .enable,
+    };
+    
+    const bg0_map: [*]volatile bg.TextScreenEntry = @ptrCast(&bg.screen_block_ram[28]);
+    
+    while (true) {
+        display.naiveVSync();
+        
+        // Convert elapsed seconds to a 2-digit display
+        const digits = bios.div(timer.data_2.*, 10);
```
```suggestion
    timer[1].initial_value = @truncate(-0x4000);
    timer[1].ctrl = .{
        .freq = .cycles_1024,
        .enable = .enable,
    };
    timer[2].ctrl = .{
        .freq = .cycles_1024,
        .cascade = .enable,
        .enable = .enable,
    };
    
    const bg0_map: [*]volatile bg.TextScreenEntry = @ptrCast(&bg.screen_block_ram[28]);
    
    while (true) {
        display.naiveVSync();
        
        // Convert elapsed seconds to a 2-digit display
        const digits = bios.div(timer[2].initial_value, 10);
```

Changes necessitated by the other suggestions.

pineapplemachine replied:
The value isn't actually time, though. Maybe "counter"? That's how the tonc documentation refers to the contents of the register.

https://gbadev.net/tonc/timers.html#ssec-reg-tmxd

![image](https://github.com/user-attachments/assets/6fd86c89-0af2-45b0-bc09-d8d6ad835ea7)

vortexofdoom commented on GBA/timer.zig:
```diff
@@ -0,0 +1,70 @@
+const std = @import("std");
+const gba = @import("gba.zig");
+const Enable = gba.utils.Enable;
+const timer = @This();
```
```suggestion
```

vortexofdoom commented on GBA/timer.zig:
```diff
@@ -0,0 +1,70 @@
+    unused_1: u3 = 0,
+    /// Raise an interrupt upon overflow.
+    interrupt: Enable = .disable,
+    /// Enable the timer.
+    enable: Enable = .disable,
+    /// Unused bits.
+    unused_2: u8 = 0,
+};
+
+/// Corresponds to tonc REG_TM1D.
+/// Writing to this register does NOT set the current timer value.
+/// It sets the INITIAL timer value for the next timer run.
+pub const data_1: *volatile u16 = @ptrFromInt(gba.mem.io + 0x100);
+
+/// Corresponds to tonc REG_TM1CNT.
+pub const ctrl_1: *volatile timer.Control = @ptrFromInt(gba.mem.io + 0x102);
+
+/// Corresponds to tonc REG_TM2D.
+/// Writing to this register does NOT set the current timer value.
+/// It sets the INITIAL timer value for the next timer run.
+pub const data_2: *volatile u16 = @ptrFromInt(gba.mem.io + 0x104);
+
+/// Corresponds to tonc REG_TM2CNT.
+pub const ctrl_2: *volatile timer.Control = @ptrFromInt(gba.mem.io + 0x106);
+
+/// Corresponds to tonc REG_TM3D.
+/// Writing to this register does NOT set the current timer value.
+/// It sets the INITIAL timer value for the next timer run.
+pub const data_3: *volatile u16 = @ptrFromInt(gba.mem.io + 0x108);
+
+/// Corresponds to tonc REG_TM3CNT.
+pub const ctrl_3: *volatile timer.Control = @ptrFromInt(gba.mem.io + 0x10a);
+
+/// Corresponds to tonc REG_TM4D.
+/// Writing to this register does NOT set the current timer value.
+/// It sets the INITIAL timer value for the next timer run.
+pub const data_4: *volatile u16 = @ptrFromInt(gba.mem.io + 0x10c);
+
+/// Corresponds to tonc REG_TM4CNT.
+pub const ctrl_4: *volatile timer.Control = @ptrFromInt(gba.mem.io + 0x10e);
```
```suggestion
pub const Timer = packed struct(u32) {
    /// Represents the data of a REG_TMxCNT timer control register.
    pub const Control = packed struct(u8) {
        /// Timer frequency.
        /// One second is equivalent to 1024 * 0x4000 clock cycles.
        freq: Frequency = .cycles_1,
        /// Cascade mode. When this bit is set, this timer will be incremented
        /// when the previous timer overflows. (The timer must also be enabled.)
        cascade: Enable = .disable,
        /// Unused bits.
        _: u3 = 0,
        /// Raise an interrupt upon overflow.
        interrupt: Enable = .disable,
        /// Enable the timer.
        enable: Enable = .disable,
    };

    /// Writing to this register does NOT set the current timer value.
    /// It sets the INITIAL timer value for the next timer run.
    initial_value: u16,
    ctrl: Timer.Control,
    _: u8,
};

/// Corresponds to tonc REG_TMx(D/CNT).
pub const timer: *volatile [4]Timer = @ptrFromInt(gba.mem.io + 0x100);
```

Wherever possible, I've been trying to convert contiguous sets of semantically related objects like this to arrays. With the more fine-grained control and confidence over layout and alignment that Zig has over old C toolchains like the ones originally used for Tonc, it seems like a natural choice to me. Combining that with consolidating the data/control fields into a single struct lets us number them both with array indices.

I've also been trying to make `_` the standard for unused fields, numbering if there's more than 1. Here, `Control` only uses 8 bits, so we can just change the underlying type and leave it to one unused field in the `Control` itself.

pineapplemachine replied:
Hi, thank you for your comments! I'll address these soon. I do want to discuss one of the suggested changes in particular, though:

I've also been working on a sound registers API (not yet ready for a PR) and in one case I noticed that the gbdev docs disagreed with the gbatek docs on whether one of the sound register bits was really unused or not, and which the tonc docs did not mention at all. (Specifically, the bit which gbatek documents as forcing 75% volume in `SOUND3CNT_H`/`REG_SND3CNT`.) Since it wasn't clear to me whether fields named `_` in Zig are actually referenceable (and will remain referenceable) e.g. per https://github.com/ziglang/zig/issues/4164 and since it seemed unlikely to know for certain whether these bits were unused and whether someone might actually want to access them anyway, I thought it was probably best to give them names like `unused_n`? Not a hill I'd die on, but I did see the approach in other structs with the `_` member and questioned if it could be improved upon.

(Sources about the disagreement on the unused bit:)

https://gbdev.gg8.se/wiki/articles/Gameboy_sound_hardware#Registers
![image](https://github.com/user-attachments/assets/fb1d3636-9d82-48dc-aeb5-744140d20974)

https://problemkaputt.de/gbatek.htm#gbasoundchannel3waveoutput
![image](https://github.com/user-attachments/assets/b09ca764-5576-4d78-b066-ddf546cc85f5)

pineapplemachine replied:
You suggested this comment by the way:

```
/// Importantly, these are *separate* registers, and assigning to them together
/// cannot be guaranteed to behave as expected.
```

But I found testing with NanoGameboyAdvance at least that this does in fact behave as expected (and I was unsure why it wouldn't?):

```zig
timer[1] = .{
        .data = @truncate(-0x4000),
        .ctrl = .{
            .freq = .cycles_1024,
            .enable = .enable,
        },
    };
```

vortexofdoom replied:
hm, on no$gba I was having it behave differently. My theory on why it would be is that the GBA only has a 16 bit data bus outside of IWRAM, and some quirk of compilation/optimization might be combining the accesses weirdly? And for an emulator, that might not lead into any weird issues since it's not *really* MMIO there. Let me see how it works on my analogue pocket.

pineapplemachine replied:
> Also, my reasoning for using `_` is expressly in the hope that they become unreferenceable.

Ok, fair enough :+1:

vortexofdoom replied:
it's little endian, so the low 8 bits of a u16 should be the same as a u8. All the edits I suggested work on no$gba for me, including making `Timer` a `packed struct(u24)` and `Control` being `packed struct(u8)`. I wonder if there are some other changes in later zig versions that might be breaking it on other emulators, are you building on latest?

pineapplemachine replied:
So far I tried building with `0.15.0-dev.643+dc6ffc28b` (very recent if not still latest), 0.14.0, and 0.14.1. All worked with NBA, none worked with no$gba or VBA. mGBA mostly didn't work but sometimes managed to display things as normal.

vortexofdoom commented on GBA/gba.zig:
```diff
@@ -13,6 +13,7 @@ pub const interrupt = @import("interrupt.zig");
 pub const math = @import("math.zig");
 pub const mem = @import("mem.zig");
 pub const obj = @import("obj.zig");
+pub const timer = @import("timer.zig");
```
```suggestion
pub const timer = @import("timer.zig").timer;
```

Just made necessary by switching to an array.

vortexofdoom reviewed (commented):
Thanks for the PR, this is a great addition! I really appreciate the helpful doc comments.  Made a few suggestions for consistency in style with the direction I've been going, but I don't want to steamroll by pushing them through.

pineapplemachine committed:
Some refactoring of new timer-related code re: code review comments

pineapplemachine committed:
Rename timer `data` to `counter`

This naming is more consistent with the terms used in tonc and gbatek documentation for these registers

pineapplemachine committed:
Rename unused bits fields in Timer.Control

vortexofdoom commented on examples/secondsTimer/secondsTimer.zig:
```diff
@@ -0,0 +1,108 @@
+    // which is the same as once per second.
+    // When it oveflows, Timer 2 will be incremented by 1 due
+    // to its "cascade" flag.
+    timer[1].counter = @truncate(-0x4000);
+    timer[1].ctrl = .{
+        .freq = .cycles_1024,
+        .enable = .enable,
+    };
```
```suggestion
    timer[1] = .{
        .counter = @truncate(-0x4000),
        .ctrl = .{
            .freq = .cycles_1024,
            .enable = .enable,
        },
    };
```

You were correct that this can work.

vortexofdoom reviewed (commented):
Another bit of feedback. I also messed around with various emulators and found that mGBA seems to need to run the BIOS to work, rather than skipping straight to the cart, and no$GBA needs to do the opposite and run straight to cart, possibly since it has a custom BIOS. Let me know if either of those change anything for you.

Also, are your changes accurately counting past 10? I think I have everything the same and for me, the 10s place is turning to 0 and the 1s place is blank. I can't immediately see why that would be, unless there's still something wonky with Zig's assembly outputs and the `bios.div` function, but in that case, I'd expect it to just not work at all.

vortexofdoom commented on examples/secondsTimer/secondsTimer.zig:
```diff
@@ -0,0 +1,108 @@
+        .freq = .cycles_1024,
+        .enable = .enable,
+    };
+    timer[2].ctrl = .{
+        .freq = .cycles_1024,
+        .cascade = .enable,
+        .enable = .enable,
+    };
```
```suggestion
    timer[2] = .{
        .ctrl = .{
            .freq = .cycles_1024,
            .cascade = .enable,
            .enable = .enable,
        },
    };
```

Looks like we need to make sure to initialize the other timer as well. I don't know if that need changes your thoughts on whether to flatten `Control` into `Timer`.

pineapplemachine replied:
The tonc example I based this on does not initialize the counter for the seconds timer, so I assumed it was unnecessary? Certainly happy to make the change though. (Note that the example uses timers 2 and 3 whereas I used timers 1 and 2. As far as I know, the choice is arbitrary unless you need timers 1 and/or 2 for sound?)

https://gbadev.net/tonc/timers.html#sec-demo

pineapplemachine committed:
Some rearranging of the new Timer struct

per review comments

pineapplemachine commented on GBA/timer.zig:
```diff
@@ -0,0 +1,52 @@
+    counter: u16,
+    
+    /// Corresponds to tonc REG_TMxCNT.
+    ctrl: Timer.Control,
```
Just push a commit for this, assuming I understood the intended change correctly? My personal preference would be for the structure to more closely reflect the underlying registers as in the former option, over the latter option, so that was the change I went with. If I've missed something here though please let me know.

pineapplemachine commented:
> Another bit of feedback. I also messed around with various emulators and found that mGBA seems to need to run the BIOS to work, rather than skipping straight to the cart, and no$GBA needs to do the opposite and run straight to cart, possibly since it has a custom BIOS. Let me know if either of those change anything for you.

Adding the same GBA BIOS I've been using with NBA in the mGBA settings and then loading builds of the examples didn't change the behavior. (Attempting to load a ROM without an existing save file plays the BIOS intro then shows the default mGBA no-rom-open graphic. Loading the same ROM again after mGBA writes a suspend state file from the first failed attempt seems to load the ROM as expected by loading that suspended state.)

I wasn't able to find any BIOS-related settings in no$gba to mess with, but as far as I know it's not loading any user-provided BIOS file.

> Also, are your changes accurately counting past 10? I think I have everything the same and for me, the 10s place is turning to 0 and the 1s place is blank. I can't immediately see why that would be, unless there's still something wonky with Zig's assembly outputs and the `bios.div` function, but in that case, I'd expect it to just not work at all.

Yes, with NBA 1.8.2 and the this build of the example, the counter works as intended past 10 seconds: https://files.pineapplemachine.com/public/temporary/secondsTimer.2025.06.01.gba

![2025-06-01 ziggba timer](https://github.com/user-attachments/assets/ff670662-28f6-426f-986e-bcf16cc3caa4)

pineapplemachine committed:
Change how timers are initialized in the example

Unclear under what conditions this is strictly necessary? Might be nice to write a comment explaining should we ever figure it out

vortexofdoom commented on GBA/timer.zig:
```diff
@@ -0,0 +1,53 @@
+/// for controlling and reading one of the GBA's four timers.
+pub const Timer = packed struct(u32) {
+    /// Enumeration of recognized timer tick frequencies.
+    pub const Frequency = enum(u2) {
+        /// One timer tick per clock cycle.
+        /// Equivalent to 1/16777216th second, or approximately 0.06 microseconds.
+        cycles_1 = 0,
+        /// One timer tick per 64 clock cycles.
+        /// Equivalent to 1/262144th second, or approximately 3.8 microseconds.
+        cycles_64 = 1,
+        /// One timer tick per 256 clock cycles.
+        /// Equivalent to 1/65536th second, or approximately 15 microseconds.
+        cycles_256 = 2,
+        /// One timer tick per 1024 clock cycles.
+        /// Equivalent to 1/16384th second, or approximately 61 microseconds.
+        cycles_1024 = 3,
+    };
+
+    /// Represents the data of a REG_TMxCNT timer control register.
+    pub const Control = packed struct(u8) {
+        /// Timer frequency.
+        /// One second is equivalent to 1024 * 0x4000 clock cycles.
+        freq: Timer.Frequency = .cycles_1,
+        /// Cascade mode. When this bit is set, this timer will be incremented
+        /// when the previous timer overflows. (The timer must also be enabled.)
+        cascade: Enable = .disable,
```
```suggestion
    ```suggestion
    pub const Frequency = enum(u3) {
        /// One timer tick per clock cycle.
        /// Equivalent to 1/16777216th second, or approximately 0.06 microseconds.
        cycles_1 = 0,
        /// One timer tick per 64 clock cycles.
        /// Equivalent to 1/262144th second, or approximately 3.8 microseconds.
        cycles_64 = 1,
        /// One timer tick per 256 clock cycles.
        /// Equivalent to 1/65536th second, or approximately 15 microseconds.
        cycles_256 = 2,
        /// One timer tick per 1024 clock cycles.
        /// Equivalent to 1/16384th second, or approximately 61 microseconds.
        cycles_1024 = 3,
        /// One timer tick every time the previous timer overflows.
        /// This has no effect on timer[0]
        cascade = 4,
    };

    /// Represents the data of a REG_TMxCNT timer control register.
    pub const Control = packed struct(u8) {
        /// Timer frequency.
        /// One second is equivalent to 1024 * 0x4000 clock cycles.
        freq: Timer.Frequency = .cycles_1,
```

I had a minor brainwave just now, I was going to suggest removing the setting of `timer[2].freq` in the example, since it's ignored when `cascade` is enabled, but then I realized that it could be encoded into the type system that setting them both was impossible (or rather, meaningless), and the cascade bit is really just tying the frequency to that of the previous timer. It does make the API slightly different from most resources, but I still think it's quite intelligible (probably even more so than thinking of them as separate), and I think it's fine to evolve into our own API.

If it had just been removal, I'd have probably just done it and merged, but once I had this idea, I wanted to run it by you first. Thanks for all your patience!

pineapplemachine replied:
I'm not actually a fan of this change! I think it's confusing API design if one sets up a cascading timer by initializing the timer frequency to a special value, and it is clearer at a look how to use this API if the cascade flag is separate. Maybe it's better to just omit the explicit `.freq` initialization in the example, if that would be less confusing? And certainly mention in a doc comment that frequency is meaningless when cascade is set.

pineapplemachine replied:
I like this idea, I think this is clearer. And no worries at all, I'm definitely figuring things out as I go with zig and GBA homebrew myself here.

As for the enable enum, I don't have a strong feeling either way. There are cases where I think an enum is very appropriate for some booleans (such as this one), and maybe the consistency of using an enum for all of them is the reasonable thing?

vortexofdoom commented on examples/secondsTimer/secondsTimer.zig:
```diff
@@ -0,0 +1,113 @@
+        },
+    };
+    timer[2] = .{
+        .counter = 0,
+        .ctrl = .{
+            .freq = .cycles_1024,
+            .cascade = .enable,
+            .enable = .enable,
+        },
```
```suggestion
        .counter = 0,
        .ctrl = .{
            .freq = .cascade,
            .enable = .enable,
        },
```

That would result in this change.

pineapplemachine commented:
You should pop in to the gbadev Discord and in particular the #music channel, where there's been some related discussion.

I'm not 100% sure but I think something funky happened recently in zig that is causing our discrepancies. I think in an earlier version of 0.14.0, the `myStruct = .{ myMember = etc }` syntax was guaranteed to initialize the new struct in a local and then assign the right-hand struct to the left-hand value in one go. But evidently the zig compiler versions I've tried (0.14.0 and later) compile this as member access, reading the left-hand struct and then mutating a member at a time. Since GBA structs do not always like being read, this has led to some misbehavior.

https://ziglang.org/download/0.14.0/release-notes.html#Decl-Literals

There is also a separate problem where apparently `__text_start` is beginning with junk instructions that cause everything except NBA to have issues loading the ROM, where I'm guessing it didn't with earlier versions of zig.

pineapplemachine committed:
Update timer API: Make the cascade flag a "mode" enum instead

pineapplemachine commented:
Oh crap. Now that I'm building with 2024.10.0-mach instead of 0.14.0+ I'm getting that issue with the counter not working correctly for 10 seconds and up. I don't think it's a timer initialization issue though, I think it's an issue with the `bios.div` call? Which is actually fixed with newer compiler versions?

pineapplemachine committed:
Update timer example to no longer use `.{}` struct assignment, per #26

https://github.com/wendigojaeger/ZigGBA/issues/26#issuecomment-2930529217
